apps/cli/src/deploy.ts

```diff
@@ -0,0 +1,458 @@
+            // create filter to read events from an AuthorityFactory contract
+            const filter = await publicClient.createContractEventFilter({
+                abi: authorityFactoryABI,
+                eventName: "AuthorityCreated",
+                address,
+                fromBlock,
+            });
+
+            // read all logs since beginning of time
+            const logs = await publicClient.getFilterLogs({ filter });
+
+            // parse logs into list of authorities
+            const authorities = logs
+                .filter((log) => log.args.authority)
+                .map((log) => log.args.authority!);
+            return authorities;
+        }
+    }
+    return [];
+};
+
+/**
+ * Check the integrity of the cartesi machine and returns its hash
+ * @param machine directory of cartesi machine snapshot to deploy
+ * @returns hash of the machine
+ */
+const check = async (machine: string): Promise<Hash> => {
+    if (!fs.existsSync(machine) || !fs.statSync(machine).isDirectory()) {
+        throw new Error(
+            "Cartesi machine snapshot not found, run 'sunodo build'",
+        );
+    }
+
+    const templateHash = fs
+        .readFileSync(path.join(machine, "hash"))
+        .toString("hex");
+    return `0x${templateHash}`;
+};
+
+const publishIPFS = async (
+    machine: string,
+    options: IPFSOptions,
+): Promise<PublishResult> => {
+    // XXX: start own IPFS node?
+
+    if (!options.url) {
+        // test if there is a local IPFS node running locally
+        const localhost = "http://127.0.0.1:5001";
+        const localhostRunning =
+            (await testConnection({ url: new URL(localhost) })) === "success";
+
+        // ask for IPFS node URL, default is local node if there is one running
+        const url = await input({
+            message: "IPFS node URL",
+            default: localhostRunning ? localhost : undefined,
+        });
+        options.url = new URL(url);
+    }
+
+    // try to connect with specified options
+    let status = await testConnection(options);
+
+    if (status === "unauthorized" && (!options.username || !options.password)) {
+        // if received unauthorized, ask for username and password
+        options.username =
+            options.username || (await input({ message: "IPFS username" }));
+        options.password =
+            options.password ||
+            (await inputPassword({ message: "IPFS password" }));
+
+        // try again, with username/password
+        status = await testConnection(options);
+    }
+
+    if (status !== "success") {
+        throw new Error(`Error connecting to IPFS ${options.url}`);
+    }
+
+    const cid = await importDirectory(machine, options);
+    return { method: "ipfs", location: cid };
+};
+
+/**
+ *
+ * @param machine directory of cartesi machine snapshot to deploy
+ * @param options
+ * @returns selected publishing method and location of published machine
+ */
+const publish = async (
```

Comment:
Not sure what's the usual (official?) term here, if publish or upload.

Reply:
Publish meaning "make it public"